File: awiki/python-client.backup.20260308/scripts/check_inbox.py
"""Check inbox, view chat history, mark messages as read.

Usage:
    # View inbox
    uv run python scripts/check_inbox.py

    # Limit result count
    uv run python scripts/check_inbox.py --limit 5

    # View chat history with a specific DID
    uv run python scripts/check_inbox.py --history "did:wba:localhost:user:abc123"

    # Mark messages as read
    uv run python scripts/check_inbox.py --mark-read msg_id_1 msg_id_2

[INPUT]: SDK (RPC calls), credential_store (load identity credentials), local_store (local persistence)
[OUTPUT]: Inbox message list / chat history / mark-read result
[POS]: Message receiving and processing script

[PROTOCOL]:
1. Update this header when logic changes
2. Check the folder's CLAUDE.md after updating
"""

import argparse
import asyncio
import json
import logging
import sys
from pathlib import Path
from typing import Any

from utils import SDKConfig, create_molt_message_client, authenticated_rpc_call, resolve_to_did

logger = logging.getLogger(__name__)
from credential_store import create_authenticator
import local_store


MESSAGE_RPC = "/message/rpc"


async def check_inbox(credential_name: str = "default", limit: int = 20) -> None:
    """View inbox."""
    config = SDKConfig()
    auth_result = create_authenticator(credential_name, config)
    if auth_result is None:
        print(f"Credential '{credential_name}' unavailable; please create an identity first")
        sys.exit(1)

    auth, data = auth_result
    async with create_molt_message_client(config) as client:
        inbox = await authenticated_rpc_call(
            client,
            MESSAGE_RPC,
            "get_inbox",
            params={"user_did": data["did"], "limit": limit},
            auth=auth,
            credential_name=credential_name,
        )

        # Store fetched messages locally (offline backfill)
        _store_inbox_messages(credential_name, data["did"], inbox)

        print(json.dumps(inbox, indent=2, ensure_ascii=False))


async def get_history(
    peer_did: str,
    credential_name: str = "default",
    limit: int = 50,
) -> None:
    """View chat history with a specific DID."""
    config = SDKConfig()
    auth_result = create_authenticator(credential_name, config)
    if auth_result is None:
        print(f"Credential '{credential_name}' unavailable; please create an identity first")
        sys.exit(1)

    auth, data = auth_result
    async with create_molt_message_client(config) as client:
        history = await authenticated_rpc_call(
            client,
            MESSAGE_RPC,
            "get_history",
            params={
                "user_did": data["did"],
                "peer_did": peer_did,
                "limit": limit,
            },
            auth=auth,
            credential_name=credential_name,
        )

        # Store fetched messages locally (offline backfill)
        _store_history_messages(credential_name, data["did"], peer_did, history)

        print(json.dumps(history, indent=2, ensure_ascii=False))


async def mark_read(
    message_ids: list[str],
    credential_name: str = "default",
) -> None:
    """Mark messages as read."""
    config = SDKConfig()
    auth_result = create_authenticator(credential_name, config)
    if auth_result is None:
        print(f"Credential '{credential_name}' unavailable; please create an identity first")
        sys.exit(1)

    auth, data = auth_result
    async with create_molt_message_client(config) as client:
        result = await authenticated_rpc_call(
            client,
            MESSAGE_RPC,
            "mark_read",
            params={
                "user_did": data["did"],
                "message_ids": message_ids,
            },
            auth=auth,
            credential_name=credential_name,
        )
        print("Marked as read successfully:")
        print(json.dumps(result, indent=2, ensure_ascii=False))


def _store_inbox_messages(
    credential_name: str, my_did: str, inbox: Any,
) -> None:
    """Store inbox messages locally (best-effort, non-critical)."""
    try:
        messages = inbox if isinstance(inbox, list) else inbox.get("messages", [])
        if not messages:
            return
        conn = local_store.get_connection()
        local_store.ensure_schema(conn)
        batch = []
        for msg in messages:
            sender_did = msg.get("sender_did", "")
            batch.append({
                "msg_id": msg.get("id", msg.get("msg_id", "")),
                "thread_id": local_store.make_thread_id(
                    my_did, peer_did=sender_did, group_id=msg.get("group_id"),
                ),
                "direction": 0,
                "sender_did": sender_did,
                "receiver_did": msg.get("receiver_did"),
                "group_id": msg.get("group_id"),
                "group_did": msg.get("group_did"),
                "content_type": msg.get("type", "text"),
                "content": str(msg.get("content", "")),
                "server_seq": msg.get("server_seq"),
                "sent_at": msg.get("sent_at") or msg.get("created_at"),
                "sender_name": msg.get("sender_name"),
            })
        local_store.store_messages_batch(conn, batch, credential_name=credential_name)
        # Record senders in contacts
        seen_dids: set[str] = set()
        for msg in messages:
            s = msg.get("sender_did", "")
            if s and s not in seen_dids:
                seen_dids.add(s)
                local_store.upsert_contact(
                    conn, did=s, name=msg.get("sender_name"),
                )
        conn.close()
    except Exception:
        logger.debug("Failed to store inbox messages locally", exc_info=True)


def _store_history_messages(
    credential_name: str, my_did: str, peer_did: str, history: Any,
) -> None:
    """Store chat history messages locally (best-effort, non-critical)."""
    try:
        messages = history if isinstance(history, list) else history.get("messages", [])
        if not messages:
            return
        conn = local_store.get_connection()
        local_store.ensure_schema(conn)
        batch = []
        for msg in messages:
            sender_did = msg.get("sender_did", "")
            is_outgoing = sender_did == my_did
            batch.append({
                "msg_id": msg.get("id", msg.get("msg_id", "")),
                "thread_id": local_store.make_thread_id(
                    my_did, peer_did=peer_did, group_id=msg.get("group_id"),
                ),
                "direction": 1 if is_outgoing else 0,
                "sender_did": sender_did,
                "receiver_did": msg.get("receiver_did"),
                "group_id": msg.get("group_id"),
                "group_did": msg.get("group_did"),
                "content_type": msg.get("type", "text"),
                "content": str(msg.get("content", "")),
                "server_seq": msg.get("server_seq"),
                "sent_at": msg.get("sent_at") or msg.get("created_at"),
                "sender_name": msg.get("sender_name"),
            })
        local_store.store_messages_batch(conn, batch, credential_name=credential_name)
        # Record senders in contacts
        seen_dids: set[str] = set()
        for msg in messages:
            s = msg.get("sender_did", "")
            if s and s not in seen_dids:
                seen_dids.add(s)
                local_store.upsert_contact(
                    conn, did=s, name=msg.get("sender_name"),
                )
        conn.close()
    except Exception:
        logger.debug("Failed to store history messages locally", exc_info=True)


def main() -> None:
    parser = argparse.ArgumentParser(description="Check inbox and manage messages")
    parser.add_argument("--history", type=str, help="View chat history with a specific DID or handle")
    parser.add_argument("--mark-read", nargs="+", type=str,
                        help="Mark specified message IDs as read")
    parser.add_argument("--limit", type=int, default=20,
                        help="Result count limit (default: 20)")
    parser.add_argument("--credential", type=str, default="default",
                        help="Credential name (default: default)")

    args = parser.parse_args()

    if args.mark_read:
        asyncio.run(mark_read(args.mark_read, args.credential))
    elif args.history:
        peer_did = asyncio.run(resolve_to_did(args.history))
        asyncio.run(get_history(peer_did, args.credential, args.limit))
    else:
        asyncio.run(check_inbox(args.credential, args.limit))


if __name__ == "__main__":
    main()

File: awiki/nodejs-client/scripts/check_inbox.js
#!/usr/bin/env node

/**
 * Check inbox, view chat history, mark messages as read.
 * 
 * Compatible with Python's check_inbox.py.
 * 
 * Usage:
 *   node scripts/check_inbox.js                    # View inbox
 *   node scripts/check_inbox.js --limit 5          # Limit result count
 *   node scripts/check_inbox.js --history did:...  # View chat history
 *   node scripts/check_inbox.js --mark-read id1,id2 # Mark messages as read
 */

import { loadIdentity } from './utils/credential_store.js';
import { createSDKConfig } from './utils/utils/config.js';
import { createMoltMessageClient } from './utils/utils/client.js';
import { authenticatedRpcCall } from './utils/utils/rpc.js';
import { resolveToDid } from './utils/utils/resolve.js';

const MESSAGE_RPC = '/message/rpc';

/**
 * View inbox.
 */
async function checkInbox(credentialName = 'default', limit = 20) {
    const config = createSDKConfig();
    const cred = loadIdentity(credentialName);
    
    if (!cred) {
        console.error(`Credential '${credentialName}' not found`);
        process.exit(1);
    }
    
    const client = createMoltMessageClient(config);
    
    try {
        const inbox = await authenticatedRpcCall(
            client,
            MESSAGE_RPC,
            'getInbox',
            { user_did: cred.did, limit },
            1,
            { auth: null, credentialName }
        );
        
        console.log('Inbox:');
        console.log(JSON.stringify(inbox, null, 2));
        
        if (inbox.messages && inbox.messages.length > 0) {
            console.log(`\nTotal: ${inbox.messages.length} message(s)`);
        }
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

/**
 * View chat history with a specific DID.
 */
async function getHistory(peerDid, credentialName = 'default', limit = 50) {
    const config = createSDKConfig();
    const cred = loadIdentity(credentialName);
    
    if (!cred) {
        console.error(`Credential '${credentialName}' not found`);
        process.exit(1);
    }
    
    // Resolve peerDid if it's a handle
    const resolvedDid = await resolveToDid(peerDid, config);
    
    const client = createMoltMessageClient(config);
    
    try {
        const history = await authenticatedRpcCall(
            client,
            MESSAGE_RPC,
            'getHistory',
            { user_did: cred.did, peer_did: resolvedDid, limit },
            1,
            { auth: null, credentialName }
        );
        
        console.log(`Chat history with ${resolvedDid}:`);
        console.log(JSON.stringify(history, null, 2));
        
        if (history.messages && history.messages.length > 0) {
            console.log(`\nTotal: ${history.messages.length} message(s)`);
        }
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

/**
 * Mark messages as read.
 */
async function markRead(messageIds, credentialName = 'default') {
    const config = createSDKConfig();
    const cred = loadIdentity(credentialName);
    
    if (!cred) {
        console.error(`Credential '${credentialName}' not found`);
        process.exit(1);
    }
    
    const client = createMoltMessageClient(config);
    
    try {
        const result = await authenticatedRpcCall(
            client,
            MESSAGE_RPC,
            'markRead',
            { user_did: cred.did, message_ids: messageIds },
            1,
            { auth: null, credentialName }
        );
        
        console.log('Messages marked as read:');
        console.log(JSON.stringify(result, null, 2));
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

/**
 * Parse command line arguments.
 */
function parseArgs() {
    const args = process.argv.slice(2);
    const result = {
        limit: 20,
        credential: 'default'
    };
    
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--limit':
                result.limit = parseInt(args[++i], 10);
                break;
            case '--history':
                result.history = args[++i];
                break;
            case '--mark-read':
                result.markRead = args[++i].split(',');
                break;
            case '--credential':
                result.credential = args[++i];
                break;
            case '--help':
            case '-h':
                printUsage();
                process.exit(0);
                break;
        }
    }
    
    return result;
}

function printUsage() {
    console.log(`
Check inbox, view chat history, mark messages as read.

Usage:
  node scripts/check_inbox.js [options]

Options:
  --limit <n>              Limit result count (default: 20)
  --history <did>          View chat history with a specific DID or handle
  --mark-read <ids>        Mark specified message IDs as read (comma-separated)
  --credential <name>      Credential name (default: default)
  --help, -h               Show this help message

Examples:
  node scripts/check_inbox.js
  node scripts/check_inbox.js --limit 5
  node scripts/check_inbox.js --history did:wba:awiki.ai:user:abc123
  node scripts/check_inbox.js --mark-read msg_id_1,msg_id_2
`);
}

// Main
const options = parseArgs();

if (options.markRead) {
    await markRead(options.markRead, options.credential);
} else if (options.history) {
    await getHistory(options.history, options.credential, options.limit);
} else {
    await checkInbox(options.credential, options.limit);
}
